### ISSUE TITLE
Kernel arguments

### ISSUE THREAD

**Issue by RobertPincus** [earth-system-radiation/rte-rrtmgp#130]
In branch `feature-reduce-in-place`, which will be merged into `develop` once OpenACC/MP issues are resolved, the kernels compute the sum over all g-points "in place" if the `fluxes` argument to the high-level Fortran driver is of type `ty_fluxes_broadband`. This is implemented in the kernels using what are essentially C-compatible optional arguments:  logical variables and kernel arguments which may not be referenced within the kernel depending on the value of the logical. So, for example, in the [call to the LW kernel](https://github.com/earth-system-radiation/rte-rrtmgp/blob/208c2923f97cd42339a3515e9e8caa24d7b65a88/rte/kernels/mo_rte_solver_kernels.F90#L261)  the arrays `broadband_up` and `_dn` (with dimensions `ncol, nlay+1`) are assigned values if the logical `do_broadband` is `.true.`, while the arrays `flux_up` and `_dn` (with dimensions `ncol, nlay+1, ngpt`) are assigned if  `do_broadband == .false.`

For the CPU versions of the kernels this has the potential advantage that the `flux_` arrays need not be allocated at their full sizes in the driver if `do_broadband == .true.`; we could rather use a small `decoy` array in the call to the kernel. This would reduce the considerable memory footprint of the calculation and was one motivation for the change. 

On the other hand at least one compiler (NAG, using debugging flags) attempts to fill every array with `intent(out)` to NaN if its values is not explicitly computed, so not providing fully-sized arrays causes a runtime error. And, on the GPU side, fluxes for each g-point are still computed in parallel, so working memory needs to be allocated even if only the sum of the spectrally-resolved fluxes are used. 

@alexeedm @Chiil I'm curious about your opinion - should we always completely allocate all the arrays passed to the kernels  even if they won't be referenced? Are there alternatives?

**Comment by m214089** [comment]
@RobertPincus The NAG compiler is exactly following the Fortran standard. You can change meanwhile some other compilers to behave similar (comply with the standard). 

However, we do need a solution for the given problem: What about assumed shape/size arrays as arguments? If one defines those as pointer and uses nullify and associated, you could mimic the optional/'present' - the array bounds can be set as local variables at the beginning of the executable statements section via the size intrinsic or ubound. 

Another option would be a BIND(C) container type containing the respective arrays. Just some unspecific thinking.

**Comment by peterukk** [comment]
Apparently optional arguments are C++ compatible under [Fortran 2018](https://stackoverflow.com/questions/40089567/calling-fortran-subroutines-with-optional-arguments-from-c/40089754#40089754), but I guess you need Fortran 2008 compatibility?

**Comment by RobertPincus** [comment]
@m214089 Agreed that the NAG behavior is the Fortran standard. 

@peterukk We could consider to update the standard although 2018 might not yet be supported by many compilers. 

The tension, it seems, is between Fortran performance, C compatibility, and flexibility. Fortran performance is best with explicit arrays (assumed-size?); C bindings (important for interoperability of kernels and front-ends) mean no optional arguments and assumed-size arrays (not assumed size, @m214089). I tried having different kernels e.g. for broadband and spectrally-detailed fluxes but a) this leads to lots of repeated code, and b) there are multiple permutations (e.g. Jacobians x broadband/spectral). 

Keep the ideas coming.

**Comment by RobertPincus** [comment]
In commit 3b5d157 on the `main` branch we resolved this issue by allocating space for all output argument, even if they are not used. This lets the code conform to the Fortran standard for assumed-size array arguments, as noted by @m214089. It means that the memory footprint can't be reduced - but at least memory traffic can be made smaller.
